efficientdet/keras/train_lib.py

```diff
@@ -0,0 +1,348 @@
+  """
+  with tf.name_scope('focal_loss'):
+    alpha = tf.convert_to_tensor(alpha, dtype=y_pred.dtype)
+    gamma = tf.convert_to_tensor(gamma, dtype=y_pred.dtype)
+
+    # compute focal loss multipliers before label smoothing, such that it will
+    # not blow up the loss.
+    pred_prob = tf.sigmoid(y_pred)
+    p_t = (y_true * pred_prob) + ((1 - y_true) * (1 - pred_prob))
+    alpha_factor = y_true * alpha + (1 - y_true) * (1 - alpha)
+    modulating_factor = (1.0 - p_t) ** gamma
+
+    # apply label smoothing for cross_entropy for each entry.
+    y_true = y_true * (1.0 - label_smoothing) + 0.5 * label_smoothing
+    ce = tf.nn.sigmoid_cross_entropy_with_logits(labels=y_true, logits=y_pred)
+
+    # compute the final loss and return
+    return alpha_factor * modulating_factor * ce / normalizer
+
+
+def _box_loss(box_outputs, box_targets, num_positives, delta=0.1):
+  """Computes box regression loss."""
+  # delta is typically around the mean value of regression target.
+  # for instances, the regression targets of 512x512 input with 6 anchors on
+  # P3-P7 pyramid is about [0.1, 0.1, 0.2, 0.2].
+  normalizer = num_positives * 4.0
+  mask = tf.not_equal(box_targets, 0.0)
+  box_loss = tf.losses.huber_loss(
+      box_targets,
+      box_outputs,
+      weights=mask,
+      delta=delta,
+      reduction=tf.losses.Reduction.SUM)
+  box_loss /= normalizer
+  return box_loss
+
+
+def _box_iou_loss(box_outputs, box_targets, num_positives, iou_loss_type):
+  """Computes box iou loss."""
+  normalizer = num_positives * 4.0
+  box_iou_loss = iou_utils.iou_loss(box_outputs, box_targets, iou_loss_type)
+  box_iou_loss = tf.reduce_sum(box_iou_loss) / normalizer
+  return box_iou_loss
+
+class DetectionLoss(tf.keras.losses.Loss):
+  """Computes total detection loss.
+
+  Computes total detection loss including box and class loss from all levels.
+  Args:
+    cls_outputs: an OrderDict with keys representing levels and values
+      representing logits in [batch_size, height, width, num_anchors].
+    box_outputs: an OrderDict with keys representing levels and values
+      representing box regression targets in [batch_size, height, width,
+      num_anchors * 4].
+    labels: the dictionary that returned from dataloader that includes
+      groundtruth targets.
+    params: the dictionary including training parameters specified in
+      default_haprams function in this file.
+
+  Returns:
+    total_loss: an integer tensor representing total loss reducing from
+      class and box losses from all levels.
+    cls_loss: an integer tensor representing total class loss.
+    box_loss: an integer tensor representing total box regression loss.
+    box_iou_loss: an integer tensor representing total box iou loss.
+  """
+  # Sum all positives in a batch for normalization and avoid zero
+  # num_positives_sum, which would lead to inf loss during training
+  def __init__(self, params, **kwargs):
+    super(DetectionLoss, self).__init__(**kwargs)
```

Comment:
We could use super() directly in py3, WDYT?

Reply:
Good point, no need to support python2 anymore. super() is good enough.